typing/typecore.ml

```diff
@@ -1165,14 +1165,13 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
             Some (p0, p, true)
         with Not_found -> None
       in
-      let candidates =
+      let constr =
         match lid.txt, constrs with
           Longident.Lident s, Some constrs when Hashtbl.mem constrs s ->
```

Comment:
I believe the guard can also be removed: in this case `s` is necessarily in the map.
(And if it wasn't, then `Env.lookup_all_constructors` wouldn't save us anyway)

Reply:
We would still need to add an `assert false` somewhere, because you don't want to get `Not_found` because of a silly bug.
Actually, this PR may be purely cosmetic. In order to fix the bug you found in #9012, this code would have to disappear. Or will we need something similar for extension constructors?

Reply:
This PR seems like an improvement on its own, regardless of the other issues that might be lying around.
So I'd suggest getting it in a good state, merging it, and that will put us on cleaner ground to think about the other issues?

I take you're point about weird error messages, I guess then we want to remove the guard and change the body to:
```
begin try Hashtbl.find constrs s
with Not_found -> assert false
```

Reply:
Nitpick: I think `assert (Hashtbl.mem constrs s)` is more readable (It looks like "I know this is in the hashtable" rather than "I don't want to think about this error condition, let's blow up") and also slightly closer to the current code.
(I agree with @trefis suggestion to make the code here as nice as possible independently of the other issues.)

Reply:
Done as @gasche suggested (less keystrokes).